src/pycea/tl/permutation.py

```diff
@@ -0,0 +1,270 @@
+        is skipped for that split.
+    keys_added
+        Attribute keys of `tdata.obst[tree].nodes` where split statistics will be stored. If `None`, `keys` are used.
+    tree
+        The `obst` key or keys of the trees to use. If `None`, all trees are used.
+    copy
+        If True, returns a :class:`DataFrame <pandas.DataFrame>` with split statistics.
+
+    Returns
+    -------
+    pd.DataFrame or None
+        If ``copy=True``, a DataFrame indexed by node identifiers with columns
+        ``f"{key_added}_split"`` and ``f"{key_added}_pval"`` for each requested key.
+        Nodes where the permutation test was skipped will have missing columns/values
+        for that key. If ``copy=False``, returns ``None`` after writing attributes
+        onto the graphs.
+    """
+
+    if isinstance(keys, str):
+        keys = [keys]
+    if keys_added is None:
+        keys_added = keys
+    if isinstance(keys_added, str):
+        keys_added = [keys_added]
+    if len(keys) != len(keys_added):
+        raise ValueError("Length of keys must match length of keys_added.")
+    if n_permutations < min_required_permutations:
+        raise ValueError("n_permutations must at least be min_required_permutations.")
+    tree_keys = tree
+    _check_tree_overlap(tdata, tree_keys)
+    trees = get_trees(tdata, tree_keys)
+
+    for key, key_added in zip(keys, keys_added):
+        data, is_array, is_square = get_keyed_obs_data(tdata, key)
+        if not(is_array or is_square):
+            data = data[key]
+        for _, t in trees.items():
+            for parent in nx.topological_sort(t):
+                children = list(t.successors(parent))
+
+                # don't do anything if not a split
+                if len(children) < 2:
+                    continue
+
+                # for each child, find all leaves
+                left_child = children[0]
+                right_child = children[1]
+
+                left_leaves = get_leaves_from_node(t, left_child)
+                right_leaves = get_leaves_from_node(t, right_child)
+
+                left_leaves_in_data = [x for x in left_leaves if x in data.index]
+                right_leaves_in_data = [x for x in right_leaves if x in data.index]
+
+                if len(left_leaves_in_data) > 0 and len(right_leaves_in_data) > 0:
+                    left_data = data.loc[left_leaves_in_data]
+                    right_data = data.loc[right_leaves_in_data]
+                else:
+                    continue
+
+                n_right = len(right_leaves_in_data)
+                n_left = len(left_leaves_in_data)
+
+                left_stat = reduction_fn(left_data.to_numpy())
+                right_stat = reduction_fn(right_data.to_numpy())
+                split_stat = difference_fn(left_stat, right_stat)
+                nx.set_edge_attributes(t, {
+                    (parent, left_child): {key_added: left_stat},
```

Comment:
I think it would make more sense to make these a node attribute. The stats are derived from the descendants of the left or right node and the parent does not influence the value. I usually use edge attributes for comparisons between parent and child nodes.